src/Microsoft.OData.Core/ODataLibraryCompatibility.cs

```diff
@@ -4,29 +4,49 @@
 // </copyright>
 //---------------------------------------------------------------------
 
-using System.Diagnostics.CodeAnalysis;
+using System;
 
 namespace Microsoft.OData
 {
     /// <summary>
     /// Library compatibility levels.
     /// </summary>
-    [SuppressMessage("Microsoft.Design", "CA1008:EnumsShouldHaveZeroValue")]
+    [Flags]
     public enum ODataLibraryCompatibility
     {
         /// <summary>
-        /// Version 6.x
+        /// No backward compatibility feature flag set.
         /// </summary>
-        Version6 = 060000,
+        None = 0,
 
         /// <summary>
-        /// Version 7.x
+        /// When enabled, the CSDL Scale and SRID "variable" value will be written "Variable" instead of "variable" for compatibility with older versions of the library.
+        /// </summary>
+        UseLegacyVariableCasing = 1 << 0,
+
+        /// <summary>
+        /// When enabled, writes a null odata annotation.@odata.null=true
+        /// </summary>
+        WriteTopLevelODataNullAnnotation = 1 << 1,
+
+        /// <summary>
+        ///  When enabled, writes '@odata.null=true' if a top-level property is single-valued and has null value.
         /// </summary>
-        Version7 = 070000,
+        WriteODataContextAnnotationForNavProperty = 1 << 2,
 
         /// <summary>
-        /// The latest version of the library.
+        /// When enabled, skips top level properties.
+        /// </summary>
+        SkipTopLevelProperty = 1 << 3,

```

Comment:
The naming here doesn't help someone understand what this flag does.
`SkipTopLevelProperty` communicates the idea that top-level properties are skipped altogether indiscriminately when this flag is enabled.

However, if you look at the context, you will realize that the flag is being used to control throwing of an exception in the following scenario:
- The property [`value is ODataNullValue || value == null`](https://github.com/OData/odata.net/blob/24ae2e335650c2533e541e1956f5b64f618ee0cd/src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertySerializer.cs#L166)
- The property is top-level - [`if (isTopLevel`](https://github.com/OData/odata.net/blob/24ae2e335650c2533e541e1956f5b64f618ee0cd/src/Microsoft.OData.Core/WriterValidator.cs#L246)
- Version7 or higher compatibility flag is enabled (or we're writing an `ODataVersion.V401` payload)

What we're effectively saying is that an exception should not be thrown if we encounter a top-level null value property when Version6 compatibility flag is enabled. We need to express this in the naming and the docstring: `DoNotThrowExceptionForTopLevelNullProperty` or `SuppressExceptionForTopLevelNullProperty`.

The usage then becomes:
```csharp
if (isTopLevel
    && (!((this.settings.LibraryCompatibility & ODataLibraryCompatibility.DoNotThrowExceptionForTopLevelNullProperty) ==ODataLibraryCompatibility.DoNotThrowExceptionForTopLevelNullProperty)) || this.settings.Version >= ODataVersion.V401))
{
    // ...
}
```

@ElizabethOkerio Please ensure that there are tests that cover all possible scenarios here.

Reply:
I think I'd prefer `DoNotThrowExceptionForTopLevelNullProperty`. "Suppress" might communicate the wrong intention